### PR TITLE
Baker - Enhancements - User controll over Baker #2592

### PR DESCRIPTION
**Pull request by bob0bob**
I did some searches to see if active work is being done on the Baker and I've suggest these features about 9 months ago as what I thought was a good enhancement.

If there is a procedure on new things and how to go about it. Please let me know. I didn't notice one but I manage there is one and hope this doesn't make people mad. Just trying to make the baker a nicer part of the colony.

This is an enhancement for the Baker. I've added a second gui that the user can select which recipes the baker will make. As you can see from the screen shots it shows everything recipe the baker can make. Click on the items you wish them to make. So you can control what they produce. If you only want cookies, just have the cookies the only item clicked. The baker will only make cookies. I've added Pumpkin Pie since it is a vanilla minecraft item.

So there are 4 recipes for the baker now.

If your town gets over loaded with food, you can also uncheck everything and the baker goes IDLE until there is an item that the baker is allowed to bake. You don't have to reassign them to another job anymore if you don't wish too.

Also, in this version, you have the ability to have one baker make cakes, and another make bread. Each building saves what is made in that building so you can have different buildings making different things.

There is no default recipe the baker will request if items are missing like before. They will find the first recipe that is allowed and request those items and until they receive those items, they will wait for the request to be fulfilled.

If you have more than one item checked for allowed recipes, the baker will just go down the list. The baker doesn't care if recipe 1 is fulfill-able and recipe 2 is not. If recipe 2 is the next item to be produce it will request those items and wait for it.

Unless, your building is higher than level 1. So the baker can make more than one item at a time. Say it was level 3. Which means they can be making 3 things at a time.

I've submitted all the code changes. I hope that you guys will accept the new changes to the baker. I think it still follows the scheme of MineColonies.

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 1]
```diff
@@ -90,6 +111,10 @@
      */
     private int ticksPassed = 0;
 
+    
+
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java]: javadoc please, what is this array. Describe it

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 2]
```diff
@@ -418,11 +469,69 @@ public void putInFurnace(final BlockPos currentFurnace, final BakingProduct baki
...
+     * update the Recipe Allowed option to the new settings of value
+     *
+     * @param pos   index pointer for array position of which Recipe to alter
+     * @param value boolean value if Recipe is Allowed to be made 
+     */
+    public void setRecipeAllowed(final int pos, final boolean value)
+    {
+    	if (pos < recipesAllowed.length)
+    	{
+    		recipesAllowed[pos] = value;
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:499]: formatting to the left

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 3]
```diff
@@ -418,11 +469,69 @@ public void putInFurnace(final BlockPos currentFurnace, final BakingProduct baki
...
      * The client view for the baker building.
      */
     public static class View extends AbstractBuildingWorker.View
     {
+
+    	/*
+    	 * list of recipes for the baker, 
+    	 * true - indicates baker is able to make recipe
+    	 */
+    	private boolean[] recipesAllowed;
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java]: you can copy this javadoc to the above array.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 4]
```diff
@@ -434,6 +543,40 @@ public View(final ColonyView c, final BlockPos l)
...
+         * @param value boolean value if Recipe is Allowed to be made 
+         * @param block block position of the building.
+         */
+        public void setRecipeAllowed(final int pos, final  boolean value,final BlockPos block)
+        {
+            MineColonies.getNetwork().sendToServer(new AssignBakerRecipeMessage(this, pos, value, block));
+        	if (pos < recipesAllowed.length)
+			{
+        		recipesAllowed[pos] = value;
+			}
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java]: formatting

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New  | Old  | Diff  |
| ------ |:-----:|:----:|:----:|:-----:|
| Total  | 34102 | 3740 | 2722 | +1018 |
| Errors |  593  |  26  |  17  |  +9   |
##### More Information:
- [Analysis - Full](https://teamcity.minecolonies.com/viewLog.html?buildId=11728&tab=Inspection&buildTypeId=bt88)

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java [hunk 5]
```diff
@@ -278,35 +293,57 @@ private AIState kneadTheDough()
...
-            return PREPARING;
+            final BakingProduct bakingProduct = new BakingProduct(storage.getPrimaryOutput().copy(), currentRecipe);
+            building.addToTasks(bakingProduct.getState(), bakingProduct);
+            currentBakingProduct = bakingProduct;
+            return getState();
+        }
+        else
+        {
+        	setDelay(UNABLE_TO_CRAFT_DELAY);
+        	return IDLE;
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java]: formatting

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java [hunk 6]
```diff
@@ -317,42 +354,58 @@ private AIState createNewProduct()
...
+	            else
+	            {
+	                copy = stack.copy();
+	            }
+	        }
+
+	        if (copy != null)
+	        {
+	            //Wheat will be reduced by chance only (Between 3 and 6- getBuildingLevel, meaning 3-5, 3-4, 3-3, 3-2, 3-1)
+	            final int form = (getOwnBuilding().getMaxBuildingLevel() + 1) - (getOwnBuilding().getBuildingLevel() + ItemStackUtils.getSize(copy));
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java]: formatting

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java [hunk 7]
```diff
@@ -317,42 +354,58 @@ private AIState createNewProduct()
...
-            {
-                req = form < 0 ? -worker.getRandom().nextInt(Math.abs(form)) : worker.getRandom().nextInt(form);
-            }
-            ItemStackUtils.changeSize(copy, req);
-            list.add(copy);
+        	final List<ItemStack> list = new ArrayList<>();
+	        for (final ItemStack stack : storage.getInput())
+	        {
+                list.add(stack);
+	        }
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java]: formatting

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java [hunk 8]
```diff
@@ -317,42 +354,58 @@ private AIState createNewProduct()
...
         {
-            if (stack.getItem() != Items.WHEAT)
-            {
-                list.add(stack);
-            }
-            else
-            {
-                copy = stack.copy();
-            }
+        	final List<ItemStack> list = new ArrayList<>();
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java]: the formatting here seems off

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34101 |  0  |  1  |  -1  |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34101 |  0  |  1  |  -1  |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

**Review by LDTTeam-Buildsystem — DISMISSED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34101 |  0  |  0  |  0   |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

**Review by LDTTeam-Buildsystem — APPROVED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34103 |  2  |  0  |  +2  |
| Errors |  593  |  0  |  0  |  +0  |
##### More Information:
- [Analysis - Full](https://teamcity.minecolonies.com/viewLog.html?buildId=12663&tab=Inspection&buildTypeId=bt88)

**Review by LDTTeam-Buildsystem — APPROVED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34103 |  0  |  0  |  0   |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 9]
```diff
@@ -14,6 +15,9 @@
 import com.minecolonies.coremod.entity.ai.citizen.baker.BakerRecipes;
 import com.minecolonies.coremod.entity.ai.citizen.baker.BakingProduct;
 import com.minecolonies.coremod.entity.ai.citizen.baker.ProductState;
+import com.minecolonies.coremod.network.messages.AssignBakerRecipeMessage;
+
+import io.netty.buffer.ByteBuf;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockFurnace;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:22]: ### Errors:
- Unused import import net.minecraft.block.BlockFurnace;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 10]
```diff
@@ -4,6 +4,7 @@
 import com.minecolonies.api.util.BlockPosUtil;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:4]: ### Errors:
- Unused import import com.minecolonies.api.util.BlockPosUtil;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 11]
```diff
@@ -14,6 +15,9 @@
 import com.minecolonies.coremod.entity.ai.citizen.baker.BakerRecipes;
 import com.minecolonies.coremod.entity.ai.citizen.baker.BakingProduct;
 import com.minecolonies.coremod.entity.ai.citizen.baker.ProductState;
+import com.minecolonies.coremod.network.messages.AssignBakerRecipeMessage;
+
+import io.netty.buffer.ByteBuf;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockFurnace;
 import net.minecraft.block.state.IBlockState;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:23]: ### Errors:
- Unused import import net.minecraft.block.state.IBlockState;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 12]
```diff
@@ -26,9 +30,16 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 import org.jetbrains.annotations.NotNull;
 
-import java.util.*;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:43]: ### Errors:
- Unused import import static com.minecolonies.api.util.constant.Constants.TICKS_SECOND;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 13]
```diff
@@ -4,6 +4,7 @@
 import com.minecolonies.api.util.BlockPosUtil;
 import com.minecolonies.api.util.Log;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:5]: ### Errors:
- Unused import import com.minecolonies.api.util.Log;

**Review by LDTTeam-Buildsystem — APPROVED** [review]
#### Analysis Complete
##### Statistics:
|        | Count | New  | Old  | Diff |
| ------ |:-----:|:----:|:----:|:----:|
| Total  | 34171 | 1763 | 1695 | +68  |
| Errors |  592  |  0   |  1   | +-1  |
##### More Information:
- [Analysis - Full](https://teamcity.minecolonies.com/viewLog.html?buildId=12689&tab=Inspection&buildTypeId=bt88)

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 14]
```diff
@@ -26,9 +30,16 @@
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:29]: ### Errors:
- Unused import import net.minecraft.init.Blocks;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 15]
```diff
@@ -26,9 +30,16 @@
 import net.minecraft.world.World;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:30]: ### Errors:
- Unused import import net.minecraft.world.World;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 16]
```diff
@@ -26,9 +30,16 @@
...
 import net.minecraftforge.common.util.Constants;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:31]: ### Errors:
- Unused import import net.minecraftforge.common.util.Constants;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 17]
```diff
@@ -26,9 +30,16 @@
...
 import net.minecraftforge.fml.common.gameevent.TickEvent;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:32]: ### Errors:
- Unused import import net.minecraftforge.fml.common.gameevent.TickEvent;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 18]
```diff
@@ -14,6 +15,9 @@
 import com.minecolonies.coremod.entity.ai.citizen.baker.BakerRecipes;
 import com.minecolonies.coremod.entity.ai.citizen.baker.BakingProduct;
 import com.minecolonies.coremod.entity.ai.citizen.baker.ProductState;
+import com.minecolonies.coremod.network.messages.AssignBakerRecipeMessage;
+
+import io.netty.buffer.ByteBuf;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:20]: ### Errors:
- Unused import import io.netty.buffer.ByteBuf;

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java [hunk 19]
```diff
@@ -14,6 +15,9 @@
...
 import net.minecraft.block.Block;
```
Comment (LDTTeam-Buildsystem) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:21]: ### Errors:
- Unused import import net.minecraft.block.Block;

Comment (bob0bob) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:21]: These import are required and will not compile without them.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java (hunk 17 above)
Comment (bob0bob) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:32]: neede import.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java (hunk 18 above)
Comment (bob0bob) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:20]: used import

**Comment by bob0bob** [comment]
Not sure what is going on with last check.  It says so many import not needed when they are.

**Comment by winex** [comment]
@LDTTeam-Buildsystem, another check please, you've gone crazy! :P

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java (hunk 7 above)
Comment (winex) [reply src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java]: yeah, every editor should SHOW tabs, `git diff` does. not to mention that those editors should not allow mixing tabs with spaces (i use tabs ONLY for indentation in my own projects).
i personally turned off using tabs (in `mcedit` from `mc`) for working on minecolonies code

**Comment by winex** [comment]
@bob0bob btw, can you check please if your new code fixes Baker not requesting **up to 3x** Wheat if he already has only **1x** Wheat in inventory?
i think there is something wrong here: https://github.com/ldtteam/minecolonies/blob/083d7de41a7a83feaa4cd80f675b18d2ba8f1c79/src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java#L323-L346
and in your code it starts at line :360
basically, he should request **at least 2x** Wheat in this case, but he doesn't!

one can confirm this by:
- clear Warehouse, Dman and Baker from Wheat
- give Baker only 4x Wheat - he'll bake 1 bread (or only 1x Wheat)
- there will be no requests... :(

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java [hunk 20]
```diff
@@ -278,35 +293,57 @@ private AIState kneadTheDough()
      */
     private AIState createNewProduct()
     {
-        progress = 0;
-        final List<IItemHandler> handlers = getOwnBuilding().getHandlers();
+    	progress = 0;
```
Comment (winex) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java:296]: formatting, think there's a TAB right before variable :(

File: src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/BakerRecipes.java [hunk 21]
```diff
@@ -56,7 +59,15 @@
...
-        final List<ItemStack> inputCake = new ArrayList<>();
+        final List<ItemStack> inputPumpkinPie = new ArrayList<>();
+        inputPumpkinPie.add(new ItemStack(Blocks.PUMPKIN.getItemDropped(null, null, 0), 1));
+//        inputPumpkinPie.add(new ItemStack(Items.PUMPKIN_SEEDS, 4));
+        inputPumpkinPie.add(new ItemStack(Items.SUGAR, 1));
+        inputPumpkinPie.add(new ItemStack(Items.EGG, 1));
+
+    	
+    	
+    	final List<ItemStack> inputCake = new ArrayList<>();
```
Comment (winex) [inline src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/BakerRecipes.java:70]: formatting, same as below

**Review by marchermans — APPROVED** [review]

**Comment by winex** [comment]
yeah, that's nice guys. but what about the 1xWheat bug? sorry for not opening new issue - i just thought discussing with latest contributor on this part is better as this code weren't merged at that time

**Comment by Raycoms** [comment]
It got fixed by this.

2018-07-19 14:52 GMT-03:00 Roman Vasiyarov <notifications@github.com>:

**Comment by winex** [comment]
@Raycoms ok, i'll test it then, thanks...

**Comment by Raycoms** [comment]
It still requests only 3 at max, but the dman brings 1 stack so he will
bring as many as he finds min 3, max 64

2018-07-19 14:55 GMT-03:00 Roman Vasiyarov <notifications@github.com>:

**Comment by bob0bob** [comment]
I know there is changes related to requesting because the baker would not keep requesting it based on status.  I did put the call check if item is request and if not request them based on the status.  So  in the past they might request it but if it gets wiped out then it would never ask again.  In the new code it calls that routine everytime when at a certain status.

So it should be working, but yeah.  The Dman will deliver a whole stack and not just what is requested.

**Comment by winex** [comment]
i see, and i'll test that to be sure that Baker doesn't get stuck as before.
about Dman - i hope if there is only 2x Wheat in the Warehouse (and he delivers it) - Baker would still be requesting at least +1x Wheat. that's how it **must** work!
+i hope that Baker wouldn't lag the server being in that _certain status_.

p.s. and, of course, thanks for the work - it's really cool! :)

**Comment by Raycoms** [comment]
Yeah, he will request until he has 3

2018-07-19 15:16 GMT-03:00 Roman Vasiyarov <notifications@github.com>:
